NullPointerException in scrapeRustcVersion

<!--
Hello and thank you for the issue!
If you would like to report a bug, we have added some points below that you can fill out.
Feel free to remove all the irrelevant text to request a new feature.
-->

## Environment

* **IntelliJ Rust plugin version:** 0.2.105.2133-192
* **Rust toolchain version:** 1.37.0, with Gentoo patches
* **IDE name and version:** IntelliJ IDEA 2019.2.2  Build #IC-192.6603.28
* **Operating system:** Gentoo Linux

## Problem description

If the commit-date from `rustc --version --verbose` does not match the regular expression, the String from `find(commitDateRe)?.groups?.get(1)?.value` renders to `null`, leaving the line `LocalDate.parse(null)`. From the JavaDoc of LocalDate.parse:

    text - the text to parse such as "2007-12-03", not null

Note the *not null* at the end. There is no `DateTimeParseException`, since the `NullPointerException` gets triggered beforehand.

This results in a state where Cargo projects can not be imported anymore, rendering the plugin useless.

## Steps to reproduce

* Install Rust on a Gentoo machine, maybe it even suffices to compile the vanilla Rust from source. It should not contain commit information.

* Ensure that `rustc --version --verbose` prints something like

        rustc 1.37.0
        binary: rustc
        commit-hash: unknown
        commit-date: unknown
        host: x86_64-unknown-linux-gnu
        release: 1.37.0
        LLVM version: 8.0

* Try to import a Cargo project in IDEA. It won't even show the toolchain version in the Settings. The log shows:

        2019-09-09 19:32:04,042 [  36561]  ERROR -    #org.rust.stdext.AsyncValue - java.lang.NullPointerException: text
        java.util.concurrent.CompletionException: java.lang.NullPointerException: text
                at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
                at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
                at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:632)
                at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
                at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088)
                at org.rust.openapiext.AsyncTasksKt$runAsyncTask$1.onThrowable(AsyncTasks.kt:38)
                at com.intellij.openapi.progress.impl.CoreProgressManager.finishTask(CoreProgressManager.java:501)
                at com.intellij.openapi.progress.impl.CoreProgressManager$4.lambda$run$0(CoreProgressManager.java:426)
                at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:312)
                at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun(LaterInvocator.java:433)
                at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:416)
                at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:399)
                at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
                at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:776)
                at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:727)
                at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
                at java.base/java.security.AccessController.doPrivileged(Native Method)
                at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
                at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:746)
                at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:878)
                at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:827)
                at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:466)
                at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:704)
                at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:465)
                at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
                at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
                at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
                at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
                at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
                at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
        Caused by: java.lang.NullPointerException: text
                at java.base/java.util.Objects.requireNonNull(Objects.java:246)
                at java.base/java.time.format.DateTimeFormatter.parse(DateTimeFormatter.java:1945)
                at java.base/java.time.LocalDate.parse(LocalDate.java:428)
                at java.base/java.time.LocalDate.parse(LocalDate.java:413)
                at org.rust.cargo.toolchain.RustToolchainKt.scrapeRustcVersion(RustToolchain.kt:150)
                at org.rust.cargo.toolchain.RustToolchainKt.access$scrapeRustcVersion(RustToolchain.kt:1)
                at org.rust.cargo.toolchain.RustToolchain.queryVersions(RustToolchain.kt:29)
                at org.rust.cargo.project.model.impl.CargoProjectImplKt$fetchRustcInfo$1.invoke(CargoProjectImpl.kt:566)
                at org.rust.cargo.project.model.impl.CargoProjectImplKt$fetchRustcInfo$1.invoke(CargoProjectImpl.kt)
                at org.rust.openapiext.AsyncTasksKt$runAsyncTask$1.run(AsyncTasks.kt:34)
                at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:894)
                at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:169)
                at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:591)
                at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:537)
                at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:59)
                at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:156)
                at com.intellij.openapi.progress.impl.CoreProgressManager$4.run(CoreProgressManager.java:408)
                at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:294)
                at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
                at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
                at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
                at java.base/java.lang.Thread.run(Thread.java:834)
        2019-09-09 19:32:04,045 [  36564]  ERROR -    #org.rust.stdext.AsyncValue - IntelliJ IDEA 2019.2.2  Build #IC-192.6603.28
        2019-09-09 19:32:04,045 [  36564]  ERROR -    #org.rust.stdext.AsyncValue - JDK: 11.0.3; VM: OpenJDK 64-Bit Server VM; Vendor: JetBrains s.r.o
        2019-09-09 19:32:04,045 [  36564]  ERROR -    #org.rust.stdext.AsyncValue - OS: Linux
        2019-09-09 19:32:04,048 [  36567]  ERROR -    #org.rust.stdext.AsyncValue - Plugin to blame: Rust version: 0.2.105.2133-192
        2019-09-09 19:32:04,048 [  36567]  ERROR -    #org.rust.stdext.AsyncValue - Last Action: Rust.AttachCargoProject

<!--
Please include as much of your codebase as needed to reproduce the error.
If the relevant files are large, please provide a link to a public repository or a [Gist](https://gist.github.com/).
-->

--- a/src/main/kotlin/org/rust/cargo/toolchain/RustToolchain.kt
+++ b/src/main/kotlin/org/rust/cargo/toolchain/RustToolchain.kt
@@ -5,6 +5,7 @@
 
 package org.rust.cargo.toolchain
 
+import com.google.common.annotations.VisibleForTesting
 import com.intellij.openapi.util.SystemInfo
 import com.intellij.openapi.util.io.FileUtil
 import com.intellij.openapi.vfs.LocalFileSystem
@@ -126,6 +127,11 @@ private fun scrapeRustcVersion(rustc: Path): RustcVersion? {
         ?.stdoutLines
         ?: return null
 
+    return parseRustcVersion(lines)
+}
+
+@VisibleForTesting
+fun parseRustcVersion(lines: List<String>): RustcVersion? {
     // We want to parse following
     //
     //  ```
@@ -147,7 +153,7 @@ private fun scrapeRustcVersion(rustc: Path): RustcVersion? {
     val versionText = releaseMatch.groups[1]?.value ?: return null
     val commitHash = find(commitHashRe)?.groups?.get(1)?.value
     val commitDate = try {
-        LocalDate.parse(find(commitDateRe)?.groups?.get(1)?.value)
+        find(commitDateRe)?.groups?.get(1)?.value?.let(LocalDate::parse)
     } catch (e: DateTimeParseException) {
         null
     }

--- a/src/test/kotlin/org/rust/cargo/toolchain/RustcVersionParsingTest.kt
+++ b/src/test/kotlin/org/rust/cargo/toolchain/RustcVersionParsingTest.kt
@@ -1,0 +1,93 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.cargo.toolchain
+
+import com.intellij.util.text.SemVer.parseFromText
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.rust.cargo.toolchain.RustChannel.*
+import kotlin.test.assertEquals
+import java.time.LocalDate.parse as parseDate
+
+@RunWith(Parameterized::class)
+class RustcVersionParsingTest(
+    private val input: String,
+    private val expectedVersion: RustcVersion?
+) {
+
+    @Test
+    fun test() {
+        val actualVersion = parseRustcVersion(input.trimIndent().lines())
+        assertEquals(expectedVersion, actualVersion)
+    }
+
+    companion object {
+        @Parameterized.Parameters
+        @JvmStatic fun data(): Collection<Array<Any>> = listOf(
+            arrayOf("""
+                rustc 1.37.0 (eae3437df 2019-08-13)
+                binary: rustc
+                commit-hash: eae3437dfe991621e8afdc82734f4a172d7ddf9b
+                commit-date: 2019-08-13
+                host: x86_64-apple-darwin
+                release: 1.37.0
+                LLVM version: 8.0    
+            """, RustcVersion(
+                parseFromText("1.37.0")!!,
+                "x86_64-apple-darwin",
+                STABLE,
+                "eae3437dfe991621e8afdc82734f4a172d7ddf9b",
+                parseDate("2019-08-13")
+            )),
+            arrayOf("""
+                rustc 1.39.0-nightly (9af17757b 2019-09-02)
+                binary: rustc
+                commit-hash: 9af17757be1cc3f672928ecf06c40a662c5ec26d
+                commit-date: 2019-09-02
+                host: x86_64-unknown-linux-gnu
+                release: 1.39.0-nightly
+                LLVM version: 9.0     
+            """, RustcVersion(
+                parseFromText("1.39.0")!!,
+                "x86_64-unknown-linux-gnu",
+                NIGHTLY,
+                "9af17757be1cc3f672928ecf06c40a662c5ec26d",
+                parseDate("2019-09-02")
+            )),
+            arrayOf("""
+                rustc 1.38.0-beta.2 (641586c1a 2019-08-21)
+                binary: rustc
+                commit-hash: 641586c1a54f1b1740f8dd796d7501e34c044da2
+                commit-date: 2019-08-21
+                host: x86_64-apple-darwin
+                release: 1.38.0-beta.2
+                LLVM version: 9.0                
+            """, RustcVersion(
+                parseFromText("1.38.0")!!,
+                "x86_64-apple-darwin",
+                BETA,
+                "641586c1a54f1b1740f8dd796d7501e34c044da2",
+                parseDate("2019-08-21")
+            )),
+            arrayOf("""
+                rustc 1.37.0
+                binary: rustc
+                commit-hash: unknown
+                commit-date: unknown
+                host: x86_64-unknown-linux-gnu
+                release: 1.37.0
+                LLVM version: 8.0                
+            """, RustcVersion(
+                parseFromText("1.37.0")!!,
+                "x86_64-unknown-linux-gnu",
+                STABLE,
+                null,
+                null
+            ))
+        )
+    }
+}

